File: alchem-link/src/sentience/perception.py
"""§2 Data/Perception  D = A_aud × Vis × X × I"""
from __future__ import annotations
from dataclasses import dataclass
from .types import Bounded


@dataclass
class Perception:
    audio: Bounded      # A_aud
    visual: Bounded     # Vis
    sensory: Bounded    # X
    integrity: Bounded  # I

    def __init__(self, audio=1.0, visual=1.0, sensory=1.0, integrity=1.0):
        self.audio = Bounded(audio)
        self.visual = Bounded(visual)
        self.sensory = Bounded(sensory)
        self.integrity = Bounded(integrity)

    def data_ratio(self) -> Bounded:
        """D = A_aud × Vis × X × I"""
        return Bounded(
            self.audio.value * self.visual.value *
            self.sensory.value * self.integrity.value
        )

    @classmethod
    def default(cls) -> "Perception":
        return cls()

File: crates/scematica-sentience/src/perception.rs
//! §2  Data / Perception Equation
//!
//! ```text
//! D = A_aud × Vis × X × I
//! ```
//!
//! Each channel is normalized to `[0,1]`.  The multiplicative form ensures that a
//! completely absent channel cannot be compensated by boosting the others.

use serde::{Deserialize, Serialize};

use crate::types::Bounded;

/// Raw sensory / perceptual inputs.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Perception {
    /// Audio perception quality (`A_aud`).
    pub audio: Bounded,
    /// Visual perception quality (`Vis`).
    pub visual: Bounded,
    /// General / environmental sensory input (`X`).
    pub sensory: Bounded,
    /// Information / data integrity (`I`).
    pub integrity: Bounded,
}

impl Perception {
    pub fn new(audio: f64, visual: f64, sensory: f64, integrity: f64) -> Self {
        Self {
            audio: audio.into(),
            visual: visual.into(),
            sensory: sensory.into(),
            integrity: integrity.into(),
        }
    }

    /// Compute the **Data Ratio** `D = A_aud × Vis × X × I`.
    pub fn data_ratio(&self) -> Bounded {
        (self.audio.value()
            * self.visual.value()
            * self.sensory.value()
            * self.integrity.value())
        .into()
    }
}

impl Default for Perception {
    /// A fully capable perception system.
    fn default() -> Self {
        Self::new(1.0, 1.0, 1.0, 1.0)
    }
}
